[WebGPU] Migrate from dataLog to LOG_WITH_STREAM

litherum opened the pull request:
#### 62446439bea635245ee820a9c1c4f0eeefa303aa
<pre>
[WebGPU] Migrate from dataLog to LOG_WITH_STREAM
<a href="https://bugs.webkit.org/show_bug.cgi?id=251831">https://bugs.webkit.org/show_bug.cgi?id=251831</a>
rdar://105111194

Reviewed by NOBODY (OOPS!).

If we migrate to LOG_WITH_STREAM, we can enable/disable log channels individually,
and have them work across processes.

* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
(WGSL::AST::dumpAST):
* Source/WebGPU/WGSL/AST/ASTStringDumper.h:
(WGSL::AST::dumpNode): Deleted.
* Source/WebGPU/WGSL/CompilationMessage.cpp: Removed.
* Source/WebGPU/WGSL/CompilationMessage.h:
(WGSL::operator&lt;&lt;):
* Source/WebGPU/WGSL/LogInitialization.h: Added.
* Source/WebGPU/WGSL/LogInitialization.mm: Added.
(WGSL::logChannels):
(WGSL::getLogChannel):
* Source/WebGPU/WGSL/Logging.h: Added.
* Source/WebGPU/WGSL/Logging.mm: Added.
(WGSL::logLevelString):
* Source/WebGPU/WGSL/Metal/MetalCodeGenerator.cpp:
(WGSL::Metal::dumpMetalCodeIfNeeded):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
* Source/WebGPU/WGSL/PhaseTimer.h:
(WGSL::dumpASTIfNeeded):
(WGSL::logPhaseTimes):
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::check):
(WGSL::TypeChecker::infer):
(WGSL::TypeChecker::resolve):
(WGSL::TypeChecker::unify):
* Source/WebGPU/WGSL/Types.h:
(WGSL::operator&lt;&lt;):
* Source/WebGPU/WGSL/WGSL.cpp:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::createFunction):
(WebGPU::createComputePipelineState):
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::GPUFrameCapture::captureFrame):
* Source/WebGPU/WebGPU/LogInitialization.h: Added.
* Source/WebGPU/WebGPU/LogInitialization.mm: Added.
(WebGPU::logChannels):
(WebGPU::getLogChannel):
* Source/WebGPU/WebGPU/Logging.h: Added.
* Source/WebGPU/WebGPU/Logging.mm: Added.
(WebGPU::logLevelString):
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::ShaderModule::createLibrary):
(WebGPU::ShaderModule::getNamedFunction const):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/62446439bea635245ee820a9c1c4f0eeefa303aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106516 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15531 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39324 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115700 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115351 "Failed to checkout and rebase branch from PR 9734") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110425 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17026 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6759 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98708 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112284 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12937 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95909 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40493 "Found 2 new test failures: imported/w3c/web-platform-tests/fetch/stale-while-revalidate/stale-css.html, imported/w3c/web-platform-tests/fetch/stale-while-revalidate/stale-script.html (failure)") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/110001 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94812 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27553 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82211 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96088 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8761 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28905 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/95532 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/6658 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9303 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5477 "Found 1 new test failure: imported/w3c/web-platform-tests/css/selectors/invalidation/lang-pseudo-class-in-has.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30648 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14919 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48453 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/104274 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10843 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25838 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebGPU/WGSL/Logging.mm:
```diff
@@ -0,0 +1,51 @@
+String logLevelString()
+{
+#if !LOG_DISABLED
+    static NSString * const defaultsDomain = @"WGSLLogging";
```
Do you want WGSL to have its own logging channels, or should the channels be defined at the WebGPU level?

litherum replied:
WGSL can't see WebGPU - it's a lower layer. So either WGSL defines logging for both WebGPU and WGSL, or I do it like I do here, and have 2 separate Logging.mm files.

litherum committed:
[WebGPU] Migrate from dataLog to LOG_WITH_STREAM
https://bugs.webkit.org/show_bug.cgi?id=251831
rdar://105111194

Reviewed by NOBODY (OOPS!).

If we migrate to LOG_WITH_STREAM, we can enable/disable log channels individually,
and have them work across processes.

* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
(WGSL::AST::dumpAST):
* Source/WebGPU/WGSL/AST/ASTStringDumper.h:
(WGSL::AST::dumpNode): Deleted.
* Source/WebGPU/WGSL/CompilationMessage.cpp: Removed.
* Source/WebGPU/WGSL/CompilationMessage.h:
(WGSL::operator<<):
* Source/WebGPU/WGSL/LogInitialization.h: Added.
* Source/WebGPU/WGSL/LogInitialization.mm: Added.
(WGSL::logChannels):
(WGSL::getLogChannel):
* Source/WebGPU/WGSL/Logging.h: Added.
* Source/WebGPU/WGSL/Logging.mm: Added.
(WGSL::logLevelString):
* Source/WebGPU/WGSL/Metal/MetalCodeGenerator.cpp:
(WGSL::Metal::dumpMetalCodeIfNeeded):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
* Source/WebGPU/WGSL/PhaseTimer.h:
(WGSL::dumpASTIfNeeded):
(WGSL::logPhaseTimes):
* Source/WebGPU/WGSL/TypeCheck.cpp:
(WGSL::TypeChecker::check):
(WGSL::TypeChecker::infer):
(WGSL::TypeChecker::resolve):
(WGSL::TypeChecker::unify):
* Source/WebGPU/WGSL/Types.h:
(WGSL::operator<<):
* Source/WebGPU/WGSL/WGSL.cpp:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::createFunction):
(WebGPU::createComputePipelineState):
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::GPUFrameCapture::captureFrame):
* Source/WebGPU/WebGPU/LogInitialization.h: Added.
* Source/WebGPU/WebGPU/LogInitialization.mm: Added.
(WebGPU::logChannels):
(WebGPU::getLogChannel):
* Source/WebGPU/WebGPU/Logging.h: Added.
* Source/WebGPU/WebGPU/Logging.mm: Added.
(WebGPU::logLevelString):
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::ShaderModule::createLibrary):
(WebGPU::ShaderModule::getNamedFunction const):

tadeuzagallo commented on Source/WebGPU/WGSL/Metal/MetalCodeGenerator.cpp:
```diff
@@ -51,8 +52,8 @@ static StringView metalCodePrologue()
     if (dumpMetalCode) {
-        dataLogLn("Generated Metal code:");
-        dataLogLn(stringBuilder.toString());
+        RELEASE_LOG_WITH_STREAM(WGSL, stream << "Generated Metal code:");
```
I don't want to bikeshed too much, but considering all these are guarded by the compile time flags, why is this better than `dataLog`? I get it for things that are always logged, but I see these as simple debugging mechanisms, that you flip a flag, recompile and get it printed to stdout. Using streams seems to require more complexity, and I honestly don't see the benefits. As far as I recall, isn't it also a little more difficult to get the information from the channel vs getting it printed to stdout?
